flink-end-to-end-tests/src/main/java/org/apache/flink/streaming/tests/StickyAllocationAndLocalRecoveryTestJob.java

```diff
@@ -0,0 +1,450 @@
+				}
+			}
+
+			// sanity check
+			if (null != valueState.value()) {
+				throw new IllegalStateException("This should never happen, keys are generated monotonously.");
+			}
+
+			// store artificial data to blow up the state
+			valueState.update(RandomStringUtils.random(valueSize, true, true));
+		}
+
+		@Override
+		public void snapshotState(FunctionSnapshotContext functionSnapshotContext) {
+		}
+
+		@Override
+		public void initializeState(FunctionInitializationContext functionInitializationContext) throws Exception {
+
+			ValueStateDescriptor<String> stateDescriptor =
+				new ValueStateDescriptor<>("state", String.class);
+			valueState = functionInitializationContext.getKeyedStateStore().getState(stateDescriptor);
+
+			ListStateDescriptor<MapperSchedulingAndFailureInfo> mapperInfoStateDescriptor =
+				new ListStateDescriptor<>("mapperState", MapperSchedulingAndFailureInfo.class);
+			schedulingAndFailureState =
+				functionInitializationContext.getOperatorStateStore().getUnionListState(mapperInfoStateDescriptor);
+
+			StreamingRuntimeContext runtimeContext = (StreamingRuntimeContext) getRuntimeContext();
+			String allocationID = runtimeContext.getAllocationID();
+
+			final int thisJvmPid = getJvmPid();
+			final Set<Integer> killedJvmPids = new HashSet<>();
+
+			// here we check if the sticky scheduling worked as expected
+			if (functionInitializationContext.isRestored()) {
+				Iterable<MapperSchedulingAndFailureInfo> iterable = schedulingAndFailureState.get();
+				String taskNameWithSubtasks = runtimeContext.getTaskNameWithSubtasks();
+
+				MapperSchedulingAndFailureInfo infoForThisTask = null;
+				List<MapperSchedulingAndFailureInfo> completeInfo = new ArrayList<>();
+				if (iterable != null) {
+					for (MapperSchedulingAndFailureInfo testInfo : iterable) {
+
+						completeInfo.add(testInfo);
+
+						if (taskNameWithSubtasks.equals(testInfo.taskNameWithSubtask)) {
+							infoForThisTask = testInfo;
+						}
+
+						if (testInfo.killedJvm) {
+							killedJvmPids.add(testInfo.jvmPid);
+						}
+					}
+				}
+
+				Preconditions.checkNotNull(infoForThisTask, "Expected to find info here.");
+
+				if (!isScheduledToCorrectAllocation(infoForThisTask, allocationID, killedJvmPids)) {
+					allocationFailureMessage = String.format(
+						"Sticky allocation test failed: Subtask %s in attempt %d was rescheduled from allocation %s " +
+							"on JVM with PID %d to unexpected allocation %s on JVM with PID %d.\n" +
+							"Complete information from before the crash: %s.",
+						runtimeContext.getTaskNameWithSubtasks(),
+						runtimeContext.getAttemptNumber(),
+						infoForThisTask.allocationId,
+						infoForThisTask.jvmPid,
+						allocationID,
+						thisJvmPid,
+						completeInfo);
+				}
+			}
+
+			// We determine which of the subtasks will produce the artificial failure
+			boolean failingTask = shouldTaskFailForThisAttempt(runtimeContext.getAttemptNumber());
+
```

Comment:
nit: since we have used the `runtimeContext` to get some task infos (`numSubtasks`, `subtaskIdx`) in `shouldTaskFailForThisAttempt()`, maybe we can also use it to get the `attemptNum` in it.

Reply:
👍